pkg/chartsync/git.go

```diff
@@ -0,0 +1,407 @@
+		_ = status.SetCondition(nsClient, *hr, status.NewCondition(
+			v1.HelmReleaseChartFetched,
+			corev1.ConditionUnknown,
+			ReasonGitNotReady,
+			"mirror not ready for sync: "+err.Error(),
+		))
+		// Repository is not ready yet; wait for signal.
+		return nil, false
+	}
+
+	// Acquire sources lock and attempt to find a source in store.
+	c.sourcesMu.Lock()
+	s, ok := c.sources[hr.ResourceID().String()]
+	if !ok {
+		// No source found in store;
+		// create the boilerplate of a new one.
+		s = &GitChartSource{Mutex: sync.Mutex{}, Mirror: mirrorName, Remote: source.GitURL}
+		c.sources[hr.ResourceID().String()] = s
+	}
+	// Acquire source lock and unlock sources lock.
+	s.Lock()
+	defer s.Unlock()
+	c.sourcesMu.Unlock()
+
+	// Get the current HEAD for the configured ref.
+	ctx, cancel := context.WithTimeout(context.Background(), c.config.GitTimeout)
+	refHead, err := repo.Revision(ctx, hr.Spec.GitChartSource.RefOrDefault())
+	cancel()
+	if err != nil {
+		logger.Log("error", "failed to get current HEAD for configured ref", "err", err.Error())
+		_ = status.SetCondition(nsClient, *hr, status.NewCondition(
+			v1.HelmReleaseChartFetched,
+			corev1.ConditionFalse,
+			ReasonGitNotReady,
+			"failed to get current HEAD for configured ref: "+err.Error(),
+		))
+		return nil, false
+	}
+
+	if ok {
+		// We did find an existing source earlier; check if the
+		// repository has seen commits in paths we are interested in.
+		ctx, cancel = context.WithTimeout(context.Background(), c.config.GitTimeout)
+		commits, err := repo.CommitsBetween(ctx, s.Head, refHead, hr.Spec.Path)
+		cancel()
+		// No commits or failed retrieving commit range,
+		// return without updating source.
+		if err != nil || len(commits) == 0 {
+			if err != nil {
+				_ = status.SetCondition(nsClient, *hr, status.NewCondition(
+					v1.HelmReleaseChartFetched,
+					corev1.ConditionFalse,
+					ReasonGitNotReady,
+					"failed to get commits between old and new revision: "+err.Error(),
+				))
+				logger.Log("error", "failed to get commits between old and new revision", "curRev", s.Head, "newRev", refHead, "err", err.Error())
+			}
+			return s, false
+		}
+	}
+
+	// Export a new working clone for the source at the recorded ref HEAD.
+	ctx, cancel = context.WithTimeout(context.Background(), c.config.GitTimeout)
+	newExport, err := repo.Export(ctx, refHead)
+	cancel()
+	if err != nil {
+		_ = status.SetCondition(nsClient, *hr, status.NewCondition(
+			v1.HelmReleaseChartFetched,
+			corev1.ConditionFalse,
+			ReasonGitNotReady,
+			"failed to clone from mirror at given revision: "+err.Error(),
+		))
+		logger.Log("error", "failed to clone from mirror at given revision", "rev", refHead, "err", err.Error())
+		// Failed creating a git clone at the given ref HEAD.
+		return s, false
+	}
+
+	if oldExport := s.Export; oldExport != nil {
+		// Defer clean-up of old export.
+		defer oldExport.Clean()
+	}
+
+	defer func() {
```

Comment:
Any reason for not simply doing this at the end of the function instead of using `defer`?

Reply:
I want to set the condition on the `HelmRelease` _after_ the source has become available so that the timestamp of the update is technically correct. Which isn't the case as long as this method holds the lock.

Reply:
We are probably talking about sub-millisecond time difference, is that really worth it? At least I think we should add a comment.